### PR TITLE
Remove wasmi backend from sc-executor

### LINKED ISSUE #13496: Remove `wasmi` as `sc-executor` backend
`wasmi` should be removed as `sc-executor` backend. It was quite useful in the early days, but it will never be as fast as `wasmtime` which is currently the de-facto standard in Substrate. For the beginning it should probably be enough to just remove the crate and keep the current structure with `sc-executor`/`sc-executor-wasmtime`. In the future we may could merge these both crates as well to remove unneeded code and to simplify the entire structure.

### PR DESCRIPTION
**Pull request by yjhmelody**
This pull request removes the `wasmi` backend. This means that the runtime can not be run with an interpreter anymore. Runtime execution will then exclusively `wasmtime`.

Closes #13496.

--- 
polkadot address: 15ouFh2SHpGbHtDPsJ6cXQfes9Cx1gEFnJJsJVqPGzBSTudr

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2629806

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2629861

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2629976

**Comment by yjhmelody** [comment]
@koute  The testing `ensure_runtime_logger_respects_host_max_log_level` is failed. But the only change so far is to change the test executor to wasmtime. It's strange.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2660862

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2660863

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2660864

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2662873

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2662874

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2662869

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2662875

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2717828

**Comment by yjhmelody** [comment]
@koute  @bkchr Please take a look

**Comment by gshep** [comment]
Hi! Looks like wasmi related code should also be remove from `sp-wasm-interface` .

**Comment by yjhmelody** [comment]
I know， but maybe downstream still use them. so i decide to remove them later

**Comment by yjhmelody** [comment]
@koute Could you review ths PR?

**Comment by koute** [comment]
@yjhmelody Yes, sorry, I just got back from vacation. I will review this PR soon.

File: bin/node/cli/benches/transaction_pool.rs [hunk 1]
```diff
@@ -24,6 +24,7 @@ use kitchensink_runtime::{constants::currency::*, BalancesCall, SudoCall};
 use node_cli::service::{create_extrinsic, fetch_nonce, FullClient, TransactionPool};
 use node_primitives::AccountId;
 use sc_client_api::execution_extensions::ExecutionStrategies;
+use sc_executor::WasmtimeInstantiationStrategy;
```
Comment (bkchr) [inline bin/node/cli/benches/transaction_pool.rs:27]: Why do we need this?

File: client/cli/src/arg_enums.rs [hunk 2]
```diff
@@ -54,17 +54,13 @@ pub const DEFAULT_WASMTIME_INSTANTIATION_STRATEGY: WasmtimeInstantiationStrategy
 #[derive(Debug, Clone, Copy, ValueEnum)]
 #[value(rename_all = "kebab-case")]
 pub enum WasmExecutionMethod {
-	/// Uses an interpreter.
-	#[clap(name = "interpreted-i-know-what-i-do")]
-	Interpreted,
```
Comment (bkchr) [inline client/cli/src/arg_enums.rs:59]: This should stay and we should instead print some warning that this doesn't exist anymore.

**Review by bkchr — APPROVED** [review]
Sorry for the delay! But it looks good!

**Review by koute — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2830422

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2830423

File: client/cli/src/arg_enums.rs [hunk 3]
```diff
@@ -76,21 +76,24 @@ pub fn execution_method_from_cli(
...
-				WasmtimeInstantiationStrategy::Pooling =>
-					sc_service::config::WasmtimeInstantiationStrategy::Pooling,
-				WasmtimeInstantiationStrategy::RecreateInstance =>
-					sc_service::config::WasmtimeInstantiationStrategy::RecreateInstance,
-				WasmtimeInstantiationStrategy::LegacyInstanceReuse =>
-					sc_service::config::WasmtimeInstantiationStrategy::LegacyInstanceReuse,
-			},
+	if let WasmExecutionMethod::Interpreted = execution_method {
+		log::warn!(
+			"Interpreted execution for wasm is deprecated. Compiled execution is used by default"
```
Comment (bkchr) [inline client/cli/src/arg_enums.rs:81]:
```suggestion
			"`interpreted-i-know-what-i-do` is deprecated and will be removed in the future. Defaults to `compiled` execution mode."
```

**Comment by michalkucharczyk** [comment]
you need to put `cumulus companion: ` into description

**Comment by bkchr** [comment]
bot merge

**Comment by bkchr** [comment]
bot rebase

**Comment by bkchr** [comment]
bot merge

**Comment by yjhmelody** [comment]
Hi @bkchr Need a retry  to merge it.

**Comment by koute** [comment]
@yjhmelody I retried the job but it still fails; looks like you have to update your companion.

**Comment by yjhmelody** [comment]
@koute The companion should be merged firstly and then no need companion PR more. Because the companion PR in cumulus not really need to update substrate commits.

**Comment by koute** [comment]
> @koute The companion should be merged firstly and then no need companion PR more. Because the companion PR in cumulus not really need to update substrate commits.

Yeah, that works too. Let's do that.

**Pull request merged**